### PR TITLE
Account-level queues

### PR DESCRIPTION
**Pull request by BrunoScheufler**
## Type of change (choose one)
- [ ] Chore (refactors, upgrades, etc.)
- [ ] Bug fix (non-breaking change that fixes an issue)
- [ ] Security fix (non-breaking change that fixes a potential vulnerability)
- [ ] Docs
- [ ] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality not to work as expected)

## Checklist
- [ ] I've linked any associated issues to this PR.
- [ ] I've tested my own changes.

*[Check our Pull Request Guidelines](https://github.com/inngest/inngest/blob/main/docs/PULL_REQUEST_GUIDELINES.md)*

### CODE REVIEW

File: pkg/execution/state/redis_state/lua/includes/update_account_queues.lua [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+    -- pointers are already second precision, so we do not want to truncate any further (as opposed to get_fn_partition_score)
+    -- earliest is a table containing {item, score}
+    return tonumber(earliestScore[2])
+end
+
+-- This function updates account queues
+-- Requires: update_pointer_score.lua, ends_with.lua
+local function update_account_queues(keyGlobalAccountPointer, keyAccountPartitions, partitionID, accountId, score)
+  -- we might be leasing an "old" partition which doesn't store the account
+  if account_is_set(keyAccountPartitions) == true then
```
Comment (BrunoScheufler) [inline pkg/execution/state/redis_state/lua/includes/update_account_queues.lua:21]: TODO: Go over this once more to make sure this is sound

PartitionLease/PartitionRequeue act on old or new partitions. Old partitions are workflow partitions, new partitions are key queues. Old partitions might not store the account ID, so the key and value will be invalid.

As soon as we enqueue a new item, the right queues should be created (as we generate new partitions by using `ItemPartitions`), and the next PartitionLease will work on that.

So this case should only happen in new executors (running the new code) that process old partitions (that haven't enqueued a new item).

File: pkg/execution/state/redis_state/queue.go [hunk 2]
```diff
@@ -1180,6 +1184,7 @@ func (q *queue) EnqueueItem(ctx context.Context, i QueueItem, at time.Time) (Que
 		parts[0].ID,
 		parts[1].ID,
 		parts[2].ID,
+		i.Data.Identifier.AccountID.String(),
```
Comment (BrunoScheufler) [inline pkg/execution/state/redis_state/queue.go:1187]: Are we sure there are no more queue items without an Identifier? We might need to check this!

File: pkg/execution/state/redis_state/lua/includes/update_account_queues.lua (hunk 1 above)
Comment (tonyhb) [reply pkg/execution/state/redis_state/lua/includes/update_account_queues.lua:21]: Correct, and that should be a safe operation as account queues are all forward-compatible;  if you're not enqueueing, you don't get them.

**Review by tonyhb — APPROVED** [review]

File: pkg/execution/state/redis_state/queue_processor.go [hunk 3]
```diff
@@ -687,6 +662,75 @@ func (q *queue) scan(ctx context.Context) error {
...
+
+	peekSize := PartitionPeekMax
+	peekUntil := q.clock.Now().Add(PartitionLookahead)
+
+	// Randomly scan account partition
+	// TODO Should we determine this for each scan iteration or determine a value at launch time?
+	// TODO We might be able to consolidate logic for handling guaranteed capacity and sequential/account/partition operation modes
+	scanAccounts := !q.isSequential() && rand.Intn(2) == 1
+	if scanAccounts {
+		peekedAccounts, err := q.accountPeek(ctx, false, peekUntil, AccountPeekMax)
```
Comment (tonyhb) [inline pkg/execution/state/redis_state/queue_processor.go:688]: Can we add this as a metric?

File: pkg/execution/state/redis_state/queue_processor.go [hunk 4]
```diff
@@ -687,6 +662,75 @@ func (q *queue) scan(ctx context.Context) error {
...
+	// By default, use the global partition
+	partitionKey := q.u.kg.GlobalPartitionIndex()
+
+	peekSize := PartitionPeekMax
+	peekUntil := q.clock.Now().Add(PartitionLookahead)
+
+	// Randomly scan account partition
+	// TODO Should we determine this for each scan iteration or determine a value at launch time?
+	// TODO We might be able to consolidate logic for handling guaranteed capacity and sequential/account/partition operation modes
+	scanAccounts := !q.isSequential() && rand.Intn(2) == 1
```
Comment (BrunoScheufler) [inline pkg/execution/state/redis_state/queue_processor.go:686]: Add a metric to track whether we're scanning accounts/partitions

**Pull request merged**